File: design-patterns/java/Strategy-DesignPattern/WithStrategyPattern/PaymentApp.java
package WithStrategyPattern;

public class PaymentApp {
    public static void main(String[] args) {
        PaymentSystem system = new PaymentSystem();

        // Scenario A: User selects Credit Card
        system.setPaymentStrategy(new CreditCardPayment());
        system.executePayment(100);

        // Scenario B: User switches to PayPal
        system.setPaymentStrategy(new PayPalPayment());
        system.executePayment(50);

        // Scenario C: User switches to BankTransfer
        system.setPaymentStrategy(new BankTransferPayment());
        system.executePayment(200);

        // Scenario D: Adding a new method (e.g., GPAY)
        // would NOT require changing PaymentSystem at all!
    }
}

File: design-patterns/python/Strategy-DesignPattern/WithStrategyPattern/payment_app.py
"""Strategy pattern demo."""
from .payment_system import PaymentSystem
from .credit_card_payment import CreditCardPayment
from .paypal_payment import PayPalPayment
from .bank_transfer_payment import BankTransferPayment


def main():
    system = PaymentSystem()

    system.set_payment_strategy(CreditCardPayment())
    system.execute_payment(100)

    system.set_payment_strategy(PayPalPayment())
    system.execute_payment(50)

    system.set_payment_strategy(BankTransferPayment())
    system.execute_payment(200)


if __name__ == "__main__":
    main()
